### ISSUE TITLE
auto-test documented test programs

### ISSUE THREAD

**Issue by markusicu** [unicode-org/unicodetools#186]
In a CI workflow, run unit tests for documented tests.
- Look through the https://github.com/unicode-org/unicodetools/tree/main/docs pages (excluding "help")
- For each test documented there,
  - Add it to the list below
  - If it is not already using JUnit, then convert it to that
  - Add it to CI

List of tests:
- [ ] TestUnicodeInvariants, see https://github.com/unicode-org/unicodetools/blob/main/docs/build.md#invariant-checking
- [ ] TODO: look for more

**Comment by markusicu** [comment]
Josh to find test programs, Steven to JUnit-ify them.

**Comment by markusicu** [comment]
Example: https://github.com/unicode-org/unicodetools/blob/main/docs/idna.md
--> add an entry in the issue description/checklist above for TestIdna.java

Use judgment; some classes that should be unit tests may not be called `*Test*.java`.

**Comment by markusicu** [comment]
Hi, just checking -- any progress? I think "Josh to find test programs" comes first...?

**Comment by josh-hadley** [comment]
Thanks for the reminder -- I have been sidetracked but hoping to make some progress on this and #187 this week

**Comment by josh-hadley** [comment]
@markusicu @srl295 I went through and added what appeared to be tests ("checks" in some cases).

**Comment by markusicu** [comment]
Thanks @josh-hadley ! Over to @srl295 ?
- TestUnicodeInvariants should be shovel-ready, and it's important because it frequently shows us needing to adjust properties.
  - Ideal would be a parameterized test, I think, but even just initially having it fail (boolean) if there is anything inconsistent would be useful.
- TestIdna is hopefully easy
- CheckXmlProperties is definitely a test, not a tool.
  - As a test, it would be useful to JUnit-ify it.
  - Issue: The ucdxml files are usually updated only rarely, maybe only once or twice per release. This means that this test would usually fail, so we can't put it into Continuous Integration.
  - Issue: The docs suggest that the test is failing already, and that one is to ignore those failures. In order to make it pass, it sounds like we would need to improve the XML file parsing.
  - Issue: Ownership of the ucdxml files & spec is up for grabs. Future uncertain.
- CheckProperties looks more like a tool than a test
  - Most of its actions print useful or interesting information, or export properties data in another format (e.g., JSON), etc.
  - "COMPARE" shows the differences between the latest data and the previous Unicode version
  - "ICU" shows the differences between the current data and the ICU data
  - None of these actions look useful in JUnit form, and especially not for Continuous Integration
  - FYI: The docs show known systematic differences for Age values

**Comment by srl295** [comment]
@markusicu  sounds good to me. I can do it along with #146 (similar work, separate PRs)

**Comment by markusicu** [comment]
@srl295 do you have time to work on TestUnicodeInvariants & TestIdna ?

**Comment by srl295** [comment]
> @srl295 do you have time to work on TestUnicodeInvariants & TestIdna ?

Generally yes maybe not today

**Comment by srl295** [comment]
I think #146 is done but need to verify

**Comment by markusicu** [comment]
friendly ping @srl295 on TestUnicodeInvariants & TestIdna :-)   (see https://github.com/unicode-org/unicodetools/issues/186#issuecomment-1061084118)

**Comment by markusicu** [comment]
It would be great to get these done soon, helping get the Unicode 15 beta ready.

**Comment by markusicu** [comment]
@srl295 back in town?

**Comment by srl295** [comment]
> @srl295 back in town?

I'll take a look tomorrow.

**Comment by markusicu** [comment]
For a start, TestUnicodeInvariants does not *have to* become a parameterized test. The text above says "consider" :-) It would already be useful to just have it run as a test case and fail when either ParseErrorCount or TestFailureCount is not zero.

**Comment by markusicu** [comment]
It looks like @srl295 turned TestIdna.java into a JUnit test already in PR #209, so I am checking that one off. (I then recently moved it from UnicodeJsps to plain unicodetools, and updated the instructions to say to run it as a JUnit test.)

**Comment by srl295** [comment]
> For a start, TestUnicodeInvariants does not *have to* become a parameterized test. The text above says "consider" :-) It would already be useful to just have it run as a test case and fail when either ParseErrorCount or TestFailureCount is not zero.

I've scheduled it for my work tomorrow.

**Comment by srl295** [comment]
> TestIdna.java

right, and it will also run along with all others when JUnit tests run.

**Comment by srl295** [comment]
@markusicu should CheckXmlProperties be JUnit-ified or not? The description and discussion seem to conflict

**Comment by markusicu** [comment]
> @markusicu should CheckXmlProperties be JUnit-ified or not? The description and discussion seem to conflict

Not. After [this discussion](https://github.com/unicode-org/unicodetools/issues/186#issuecomment-1061084118) we decided to skip it, and removed the check box from the description.

I will propose a small docs update before closing this issue.

**Comment by srl295** [comment]
🎉
